### PR TITLE
TEAMNADO-3212 list activation keys in the red hat connector dashboard

### PR DESCRIPTION
**Pull request by seodma**
This adds a activation keys table to the app. A generic error page is shown if the table fails to load.
For local testing the app can be run with `PROXY=true npm run start:beta`
 and then visited at https://stage.foo.redhat.com:1337/beta/settings/connector/activation-keys

### CODE REVIEW

File: profiles/local-frontend.js [hunk 1]
```diff
@@ -3,13 +3,13 @@ const APP_ID = 'connector';
 const FRONTEND_PORT = 8002;
 const routes = {};
 
-routes[`/beta/${SECTION}/${APP_ID}`] = {
+routes[`beta/${SECTION}/${APP_ID}`] = {
```
Comment (zanewoodfin) [inline profiles/local-frontend.js]: Why did you remove this `/`?  It's in both of our apps.

File: profiles/local-frontend.js [hunk 2]
```diff
@@ -3,13 +3,13 @@ const APP_ID = 'connector';
...
   host: `https://localhost:${FRONTEND_PORT}`,
 };
 routes[`/${SECTION}/${APP_ID}`] = {
   host: `https://localhost:${FRONTEND_PORT}`,
 };
 routes[`/beta/apps/${APP_ID}`] = { host: `https://localhost:${FRONTEND_PORT}` };
 routes[`/apps/${APP_ID}`] = { host: `https://localhost:${FRONTEND_PORT}` };
-
+routes[`/beta/config`] = { host: `http://localhost:8889` };
```
Comment (zanewoodfin) [inline profiles/local-frontend.js]: This seems like it shouldn't be here, and the app runs without it.  Does it serve a purpose I don't see?

File: src/Components/ActivationKeysTable/ActivationKeysTable.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  };
+  const { isLoading, error, data } = useActivationKeys();
+
+  if (isLoading && !error) {
+    return <Loading />;
+  } else if (!isLoading && !error) {
+    return (
+      <TableComposable
+        aria-label="ActivationKeys"
+        ouiaId={uuid()}
```
Comment (zanewoodfin) [inline src/Components/ActivationKeysTable/ActivationKeysTable.js]: ouiaId and ouiaSafe should be generated automatically.  I just added them to our other app as a hack because I couldn't figure out why the table wasn't working.  Did you check whether the page works without those specified anywhere or do you also need them for this page to work?

File: src/Components/ActivationKeysTable/ActivationKeysTable.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            <Th>{columnNames.usage}</Th>
+          </Tr>
+        </Thead>
+        <Tbody>
+          {data.map((datum) => (
+            <Tr key={datum.name} ouiaId={uuid()} ouiaSafe={true}>
+              <Td dataLabel={columnNames.name}>{datum.name}</Td>
+              <Td dataLabel={columnNames.role}>{datum.role}</Td>
+              <Td dataLabel={columnNames.role}>{datum.serviceLevel}</Td>
+              <Td dataLabel={columnNames.role}>{datum.role}</Td>
```
Comment (zanewoodfin) [inline src/Components/ActivationKeysTable/ActivationKeysTable.js]: Looks like there are a few copy/paste issues.  This should have `datum.usage`, right?  Also this row and the one above it need their `dataLabel` updated.

File: src/Components/ActivationKeysTable/ActivationKeysTable.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+} from '@patternfly/react-table';
+import { v4 as uuid } from 'uuid';
+import useActivationKeys from '../../hooks/useActivationKeys';
+import Loading from '../LoadingState/Loading';
+
+const ActivationKeysTable = () => {
+  const columnNames = {
+    name: 'Key Name',
+    role: 'Role',
+    sla: 'SLA',
```
Comment (zanewoodfin) [inline src/Components/ActivationKeysTable/ActivationKeysTable.js]: This key may as well be `serviceLevel` to match the response data key.

File: src/Components/ActivationKeysTable/ActivationKeysTable.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    usage: 'Usage',
+  };
+  const { isLoading, error, data } = useActivationKeys();
+
+  if (isLoading && !error) {
+    return <Loading />;
+  } else if (!isLoading && !error) {
+    return (
+      <TableComposable
```
Comment (zanewoodfin) [inline src/Components/ActivationKeysTable/ActivationKeysTable.js]: The surrounding conditional statement will be more readable if you pull this TableComposable component out into a custom component that you can define above.  In Subscription Inventory, I just defined a Results component; you can see that [here](https://github.com/RedHatInsights/subscription-inventory-ui/blob/44b24d6e84aeab975f4457fb8aa8888aa67ea9b6/src/components/ProductsTable/ProductsTable.tsx#L93).

File: src/Components/ActivationKeysTable/__tests__/__snapshots__/ActivationKeysTable.test.js.snap [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        <td
+          class=""
+          data-label="Role"
+        />
+        <td
+          class=""
+          data-label="Role"
+        >
+          B
+        </td>
```
Comment (zanewoodfin) [inline src/Components/ActivationKeysTable/__tests__/__snapshots__/ActivationKeysTable.test.js.snap:81]: These last two td tags should have C and D as their content in this snapshot, so there are a couple problems somewhere.

File: src/Components/ActivationKeysTable/ActivationKeysTable.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+              <Td dataLabel={columnNames.role}>{datum.role}</Td>
+              <Td dataLabel={columnNames.role}>{datum.serviceLevel}</Td>
+              <Td dataLabel={columnNames.role}>{datum.role}</Td>
+            </Tr>
+          ))}
+        </Tbody>
+      </TableComposable>
+    );
+  } else {
+    return <h1>Unavailable</h1>;
```
Comment (zanewoodfin) [inline src/Components/ActivationKeysTable/ActivationKeysTable.js]: This should probably render our actual shared Unavailable component.  See [here](https://github.com/RedHatInsights/subscription-inventory-ui/blob/44b24d6e84aeab975f4457fb8aa8888aa67ea9b6/src/components/ProductsTable/ProductsTable.tsx#L134).

File: src/App.js [hunk 9]
```diff
@@ -25,10 +37,12 @@ const App = () => {
   }, []);
 
   return (
-    <Fragment>
-      <NotificationsPortal />
-      <Routes />
-    </Fragment>
+    <QueryClientProvider client={queryClient}>
+      <Fragment>
```
Comment (zanewoodfin) [inline src/App.js]: I think you can safely remove this `Fragment` wrapper.  Components can only return a single value.  I don't think Fragments do anything besides bundle multiple return values together into one, but your QueryClientProvider is now doing that.

File: src/Components/ActivationKeys/index.js [hunk 10]
```diff
@@ -1,31 +1,54 @@
...
+    return (
+      <React.Fragment>
+        <PageHeader>
+          <PageHeaderTitle
+            title={
+              <Stack>
+                <StackItem>Activation Keys</StackItem>
+                <StackItem isFilled>
+                  <TextContent>
+                    <Text component={TextVariants.h6}>
```
Comment (zanewoodfin) [inline src/Components/ActivationKeys/index.js]: It's bad practice to choose h tags based on their size, rather than their hierarchy in the page.  I don't think this is logically a heading at all, so it should probably just be a `TextVariants.p`.

File: src/Components/ActivationKeys/index.js [hunk 11]
```diff
@@ -1,31 +1,54 @@
...
+
+const ActivationKeys = () => {
+  const queryClient = useQueryClient();
+  const user = queryClient.getQueryData('user');
+  const Page = () => {
+    return (
+      <React.Fragment>
+        <PageHeader>
+          <PageHeaderTitle
+            title={
```
Comment (zanewoodfin) [inline src/Components/ActivationKeys/index.js]: The `title` prop should only contain the "Activation Keys" text.  Then the org id stuff would be in a component beneath the `PageHeaderTitle` component.  Look [here](https://github.com/RedHatInsights/subscription-central-ui/blob/29ef6e72d93d2f0c0b45515e447750e35e10c4ef/src/pages/SatelliteManifestPage/SatelliteManifestPage.tsx#L24) for an example.  (We shouldn't have used a 'p' tag in that example.  That must have been missed in code review, but otherwise the layout is good.)  Then you won't need any of this `Stack` stuff.

File: src/utils/platformServices.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+const authenticateUser = () => {
```
Comment (zanewoodfin) [inline src/utils/platformServices.js:1]: Add tests for this method.  Examples [here](https://github.com/RedHatInsights/subscription-inventory-ui/blob/44b24d6e84aeab975f4457fb8aa8888aa67ea9b6/src/utilities/__tests__/platformServices.test.ts#L19).

File: src/Components/ActivationKeys/index.js [hunk 13]
```diff
@@ -1,31 +1,54 @@
 import React from 'react';
```
Comment (zanewoodfin) [inline src/Components/ActivationKeys/index.js]: Could you add tests for this file?  Without having looked too carefully, I'd assume the tests would be similar to the ones [here](https://github.com/RedHatInsights/subscription-inventory-ui/blob/main/src/pages/SubscriptionInventoryPage/__tests__/SubscriptionInventoryPage.test.tsx).

File: src/Components/LoadingState/Loading.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
```
Comment (zanewoodfin) [inline src/Components/LoadingState/Loading.js:1]: This file is technically getting coverage from another file's test, but it could use its own test like [here](https://github.com/RedHatInsights/subscription-inventory-ui/blob/main/src/components/emptyState/__tests__/Processing.test.tsx).

**Review by zanewoodfin — CHANGES_REQUESTED** [review]

File: src/Components/ActivationKeys/index.js [hunk 15]
```diff
@@ -1,31 +1,36 @@
...
-);
+const ActivationKeys = () => {
+  const queryClient = useQueryClient();
+  const user = queryClient.getQueryData('user');
+  const Page = () => {
+    return (
+      <React.Fragment>
+        <PageHeader>
+          <PageHeaderTitle title="Activation Keys" />
+          <TextVariants.p>Orginazaion ID: {user.orgId}</TextVariants.p>
```
Comment (zanewoodfin) [inline src/Components/ActivationKeys/index.js:20]: `TextVariants.p` is just supposed to be passed as a prop to the Text component, which is supposed to be nested in a TextContent component.  The patternfly docs for it are [here](https://www.patternfly.org/v4/components/text).

File: src/Components/ActivationKeys/index.js (hunk 10 above)
Comment (zanewoodfin) [reply src/Components/ActivationKeys/index.js]: `TextVariants.p` is just supposed to be passed as a prop to the Text component, which is supposed to be nested in a TextContent component.  The patternfly docs for it are [here](https://www.patternfly.org/v4/components/text).

File: src/Components/ActivationKeys/index.js (hunk 13 above)
Comment (zanewoodfin) [reply src/Components/ActivationKeys/index.js]: Seeing the tests at ActivationKeys.test.js made me think the ActivationKeys component should probably be moved out to its own file to match everything else.  Then this index.js file can just be used for exports like our other index.js files.

File: src/Components/ActivationKeys/ActivationKeys.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+const ActivationKeys = () => {
+  const queryClient = useQueryClient();
+  const user = queryClient.getQueryData('user');
+  const Page = () => {
+    return (
+      <React.Fragment>
+        <PageHeader>
+          <PageHeaderTitle title="Activation Keys" />
+          <Text component={TextVariants.p}>Orginazaion ID: {user.orgId}</Text>
```
Comment (zanewoodfin) [inline src/Components/ActivationKeys/ActivationKeys.js]: I think this component is also supposed to be contained in a TextContent component as seen [here](https://www.patternfly.org/v4/components/text#body).

**Review by zanewoodfin — APPROVED** [review]

File: src/Components/ActivationKeys/__tests__/ActivationKeys.test.js [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+  <div>Activation Keys Table</div>
+));
+// eslint-disable-next-line react/display-name
+jest.mock('../no-access', () => () => <div>Not Authorized</div>);
+
+jest.mock(
+  '@redhat-cloud-services/frontend-components/Unavailable',
+  // eslint-disable-next-line react/display-name
+  () => () => <div>Unavailable</div>
+);
```
Comment (bastilian) [inline src/Components/ActivationKeys/__tests__/ActivationKeys.test.js:63]: Is there a reason why these are mocked?

File: src/Components/ActivationKeysTable/ActivationKeysTable.js [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+const ActivationKeysTable = () => {
+  const columnNames = {
+    name: 'Key Name',
+    role: 'Role',
+    serviceLevel: 'SLA',
+    usage: 'Usage',
+  };
+  const { isLoading, error, data } = useActivationKeys();
+
+  const Results = () => {
```
Comment (bastilian) [inline src/Components/ActivationKeysTable/ActivationKeysTable.js:23]: I would try to avoid defining components in components, as it does degrade rendering performance a bit.
I'd rather define the component outside and pass in the `data` as a prop.

File: package.json [hunk 19]
```diff
@@ -72,23 +73,27 @@
     "@wojtekmaj/enzyme-adapter-react-17": "^0.4.1",
     "axios-mock-adapter": "^1.19.0",
     "babel-core": "7.0.0-bridge.0",
-    "babel-jest": "26.6.3",
+    "babel-jest": "=26.6.3",
```
Comment (bastilian) [inline package.json:76]: Is this pinned on purpose?

File: package.json [hunk 20]
```diff
@@ -72,23 +73,27 @@
...
+    "bdd-lazy-var": "^2.6.1",
     "cross-fetch": "^3.0.6",
     "css-loader": "5.0.1",
     "enzyme": "3.11.0",
     "enzyme-to-json": "3.6.1",
     "eslint": "7.19.0",
     "eslint-loader": "4.0.2",
     "identity-obj-proxy": "3.0.0",
-    "jest": "26.6.3",
+    "jest": "=26.6.3",
```
Comment (bastilian) [inline package.json:88]: Same as above.

File: src/hooks/__tests__/useRbacPermissions.test.js [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    {
+      resourceDefinitions: [],
+      permission: 'config-manager:activation_keys:read',
+    },
+  ];
+
+  let rbacObject = { canReadActivationKeys: true };
+
+  beforeEach(() => {
+    Object.defineProperty(window, 'insights', {
```
Comment (bastilian) [inline src/hooks/__tests__/useRbacPermissions.test.js:20]: Why not just a simple assigning via `window.insights =`?

**Review by bastilian — APPROVED** [review]
LGTM! I couldn't test it functionally, but code wise looks good to me! Only a few minor things I was curious about, but nothing hindering merging this.

Thank you @seodma!
